File: src/java/ejercicios/Ejercicio10.java
package ejercicios;

import java.util.LinkedList;
import java.util.Queue;
import java.util.concurrent.Semaphore;
import java.util.concurrent.atomic.AtomicInteger;

public class Ejercicio10 {
    public static void main(String[] args) throws InterruptedException {
        // 1. Mutex / Lock - proteger variable compartida
        System.out.println("=== 1. Mutex / Lock ===");
        Object lock = new Object();
        int[] contador = {0};
        Thread[] hilos = new Thread[5];
        for (int i = 0; i < 5; i++) {
            hilos[i] = new Thread(() -> {
                for (int j = 0; j < 1000; j++) {
                    synchronized (lock) { contador[0]++; }
                }
            });
            hilos[i].start();
        }
        for (Thread h : hilos) h.join();
        System.out.println("Contador con lock: " + contador[0] + " (esperado: 5000)");

        // 2. Semaphore - solo 2 hilos a la vez
        System.out.println("\n=== 2. Semaphore (max 2 accesos simultaneos) ===");
        Semaphore semaforo = new Semaphore(2);
        for (int i = 0; i < 5; i++) {
            int id = i;
            new Thread(() -> {
                try {
                    semaforo.acquire();
                    System.out.println("Hilo " + id + " accedio al recurso");
                    Thread.sleep(200);
                    semaforo.release();
                } catch (InterruptedException e) { }
            }).start();
        }
        pausa(600);

        // 3. Monitor / Condition - productor-consumidor
        System.out.println("\n=== 3. Monitor / Condition ===");
        Queue<Integer> cola = new LinkedList<>();
        Object monLock = new Object();
        Thread productor = new Thread(() -> {
            for (int i = 1; i <= 5; i++) {
                synchronized (monLock) {
                    cola.add(i);
                    System.out.println("Producido: " + i);
                    monLock.notify();
                }
                pausa(100);
            }
        });
        Thread consumidor = new Thread(() -> {
            for (int i = 0; i < 5; i++) {
                synchronized (monLock) {
                    while (cola.isEmpty()) {
                        try { monLock.wait(); } catch (InterruptedException e) { }
                    }
                    System.out.println("Consumido: " + cola.poll());
                }
                pausa(200);
            }
        });
        productor.start();
        consumidor.start();
        productor.join();
        consumidor.join();

        // 4. Atomic operations
        System.out.println("\n=== 4. Atomic operations ===");
        AtomicInteger atomic = new AtomicInteger(0);
        for (int i = 0; i < 5; i++) {
            hilos[i] = new Thread(() -> {
                for (int j = 0; j < 1000; j++) atomic.incrementAndGet();
            });
            hilos[i].start();
        }
        for (Thread h : hilos) h.join();
        System.out.println("AtomicInteger: " + atomic.get() + " (esperado: 5000)");

        // 5. Critical Section
        System.out.println("\n=== 5. Critical Section ===");
        System.out.println("Seccion critica: codigo dentro de synchronized { }");
        System.out.println("Solo un hilo ejecuta esa seccion a la vez.");
    }

    private static void pausa(int ms) {
        try { Thread.sleep(ms); } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }
}

File: src/python/ejercicios/Ejercicio10.py
import threading
import time
from collections import deque

# 1. Mutex / Lock
print("=== 1. Mutex / Lock ===")
lock = threading.Lock()
contador = 0

def incrementar():
    global contador
    for _ in range(1000):
        with lock:
            contador += 1

hilos = [threading.Thread(target=incrementar) for _ in range(5)]
for h in hilos: h.start()
for h in hilos: h.join()
print(f"Contador con lock: {contador} (esperado: 5000)")

# 2. Semaphore
print("\n=== 2. Semaphore (max 2 accesos simultaneos) ===")
semaforo = threading.Semaphore(2)

def acceder(id):
    with semaforo:
        print(f"Hilo {id} accedio al recurso")
        time.sleep(0.2)

hilos = [threading.Thread(target=acceder, args=(i,)) for i in range(5)]
for h in hilos: h.start()
for h in hilos: h.join()

# 3. Monitor / Condition
print("\n=== 3. Monitor / Condition ===")
cola = deque()
cond = threading.Condition()

def productor():
    for i in range(1, 6):
        with cond:
            cola.append(i)
            print(f"Producido: {i}")
            cond.notify()
        time.sleep(0.1)

def consumidor():
    for _ in range(5):
        with cond:
            while not cola:
                cond.wait()
            print(f"Consumido: {cola.popleft()}")
        time.sleep(0.2)

p = threading.Thread(target=productor)
c = threading.Thread(target=consumidor)
p.start()
c.start()
p.join()
c.join()

# 4. Atomic operations (simulado con Lock)
print("\n=== 4. Atomic operations ===")
contador_atomico = 0
lock_atomico = threading.Lock()

def incrementar_atomico():
    global contador_atomico
    for _ in range(1000):
        with lock_atomico:
            contador_atomico += 1

hilos = [threading.Thread(target=incrementar_atomico) for _ in range(5)]
for h in hilos: h.start()
for h in hilos: h.join()
print(f"Contador atomico: {contador_atomico} (esperado: 5000)")

# 5. Critical Section
print("\n=== 5. Critical Section ===")
print("Seccion critica: codigo dentro de with lock:")
print("Solo un hilo ejecuta esa seccion a la vez.")
